notebooks/agentic_rag_pipeline_with_memory_and_chat.ipynb

```diff
@@ -0,0 +1,530 @@
+      ],
+      "metadata": {
+        "id": "Er_i5ClSaCb6"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "from haystack.components.retrievers import FilterRetriever\n",
+        "from haystack.components.writers import DocumentWriter\n",
+        "from haystack import Document\n",
+        "from haystack.components.converters import OutputAdapter\n",
+        "from typing import List\n",
+        "from haystack.document_stores.types import DuplicatePolicy\n",
+        "\n",
+        "memory_store = InMemoryDocumentStore()\n",
+        "memory_retriever = FilterRetriever(memory_store)\n",
+        "# The same ChatMessage can't be stored multiple times in memory_store\n",
+        "memory_writer = DocumentWriter(memory_store, policy=DuplicatePolicy.SKIP)"
+      ],
+      "metadata": {
+        "id": "KxLgfi11Z3BT"
+      },
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "## Save `ChatMessage` objects\n",
+        "\n",
+        "As the pipeline runs through multiple loops, you'll need to save previous chat messages to the document store.\n",
+        "\n",
+        "This is a utility function that takes a list of `ChatMessage` objects and turns them into Haystack documents. The `OutputAdapter` component makes the `chat_messages_to_docs` function fit smoothly into our pipeline, passing the input from the component at the end of the previous loop iteration into the next."
+      ],
+      "metadata": {
+        "id": "jF4lvfzdbkvz"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "def chat_messages_to_docs(chat_messages: List[ChatMessage]):\n",
+        "    return [Document(content=message.content) for message in chat_messages]\n",
+        "\n",
+        "output_adapter = OutputAdapter(template=\"{{ chat_messages | chat_messages_to_docs }}\", output_type=List[Document], custom_filters={\"chat_messages_to_docs\": chat_messages_to_docs})"
+      ],
+      "metadata": {
+        "id": "7hq8rZMvbrBz"
+      },
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "## Create the prompt\n",
+        "\n",
+        "The system message tells the LLM how to act.\n",
+        "\n",
+        "We need 2 loops here in our prompt: one to run through the memories, one to run through any relevant documents that were returned."
+      ],
+      "metadata": {
+        "id": "PAREbRszj-29"
+      }
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "For simplicity, this pipelines uses the `InMemoryBM25Retriever` to do keyword-based retrieval. In a production scenario where you're searching a vast number of documents, an `EmbeddingRetriever` would be faster and more accurate.\n",
```

Comment:
Let's not say that an `EmbeddingRetriever` would be faster. More accurate is enough.